src/runtime/algorithm.go

```diff
@@ -0,0 +1,55 @@
+package runtime
+
+// This file implements various core algorithms used in the runtime package and
+// standard library.
+
+import "unsafe"
+
+// This function is used by hash/maphash.
+func fastrand() uint32 {
+	xorshift32State = xorshift32(xorshift32State)
+	return xorshift32State
+}
+
+var xorshift32State uint32 = 1
+
+func xorshift32(x uint32) uint32 {
+	// Algorithm "xor" from p. 4 of Marsaglia, "Xorshift RNGs".
+	// This could probably be improved if needed.
+	x ^= x << 13
```

Comment:
Thanks! I have updated the code. The (7, 1, 9) constants are also better on AVR: they're much easier to shift than (13, 17, 5) because AVR doesn't have a barrel shifter.

Note sure how to best cite the algorithm, suggestions welcome :)